8.1.2/startup/generateStartupCommand.js

```diff
@@ -18,15 +18,45 @@ if (typeof process.env.WEBSITE_ROLE_INSTANCE_ID !== 'undefined'
     roleInstanceId = process.env.WEBSITE_ROLE_INSTANCE_ID;
 }
 
-var startupCommand = fs.readFileSync(CMDFILE, 'utf8').trim();
+// Is Application Insights enabled, with an associated ikey?
+var appInsightsEnabled = process.env.ENABLE_APPINSIGHTS && process.env.APPINSIGHTS_INSTRUMENTATIONKEY;
+var appInsightsPreloadArg = "--require /opt/startup/initAppInsights.js";
+var nodeCommandPrefix = "node ";
+
+if (appInsightsEnabled) {
+    console.log("Application Insights enabled");
+    nodeCommandPrefix += appInsightsPreloadArg + " ";
+}
+
+function augmentCommandIfNeccessary(command) {
+    if (!command || !appInsightsEnabled) {
+        return command;
+    }
+
+    // Application Insights is enabled, so we need to
+    // to update the specified startup command to pre-load it.
+    if (command.indexOf("pm2 start ") === 0) {
+        return command += " --node-args='" + appInsightsPreloadArg + "'";
+    } else if (command.indexOf("node ") === 0) {
+        // Simply replacing the prefix allows the user to specify
+        // additional Node flags, in addition to the AI preload one.
+        return command.replace("node ", nodeCommandPrefix);
+    }
+
+    // The command is using an unknown executable, and therefore,
+    // the App Insights runtime can't be automatically enabled.
+    return command;
+}
+
+var startupCommand = augmentCommandIfNeccessary(fs.readFileSync(CMDFILE, 'utf8').trim());
 
 // No user-provided startup command, check for scripts.start
 if (!startupCommand) {
     var packageJsonPath = "/home/site/wwwroot/package.json";
     var json = fs.existsSync(packageJsonPath) && JSON.parse(fs.readFileSync(packageJsonPath, 'utf8'))
     if (typeof json == 'object' && typeof json.scripts == 'object' && typeof json.scripts.start == 'string') {
-        console.log("Found scripts.start in package.json")
-        startupCommand = 'npm start';
+        console.log("Found scripts.start in package.json");
+        startupCommand = augmentCommandIfNeccessary(json.scripts.start.trim());
```

Comment:
After thinking about this further, and looking at a set of example apps that have `start` scripts, I think this logic is sufficient.